### PR TITLE
Adds new cpu backend without openblas for users who want to deploy on embedded platforms

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
Removes openblas as a mandatory dependency for cpu and introduces a new nd4j-minimizer backend.
This backend is for people on mobile and edge devices who want to just leverage a wrapper for a self contained c++ library.

This can be used in combination with  the custom datatype/ops to create a minimal nd4j backend.

Note this is only for cpu right now. Cuda and alternative devices would need more work.
(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [ X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ X] Created tests for any significant new code additions.
- [ X] Relevant tests for your changes are passing.

### CODE REVIEW

File: deeplearning4j/deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/iterator/SparkAMDSI.java [hunk 1]
```diff
@@ -106,7 +105,6 @@ public SparkAMDSI(MultiDataSetIterator iterator, int queueSize, BlockingQueue<Mu
 
     @Override
     protected void externalCall() {
-        TaskContextHelper.setTaskContext(context);
```
Comment (treo) [inline deeplearning4j/deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/iterator/SparkAMDSI.java:109]: This looks a bit suspicious. Why should externalCall be a noop here?

File: deeplearning4j/deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/iterator/SparkADSI.java [hunk 2]
```diff
@@ -108,7 +107,6 @@ public SparkADSI(DataSetIterator iterator, int queueSize, BlockingQueue<DataSet>
 
     @Override
     protected void externalCall() {
-        TaskContextHelper.setTaskContext(context);
```
Comment (treo) [inline deeplearning4j/deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/iterator/SparkADSI.java:111]: This looks a bit suspicious. Why should externalCall be a noop here?

File: nd4j/nd4j-backends/nd4j-backend-impls/nd4j-minimizer/src/main/resources/META-INF/services/org.nd4j.linalg.compression.NDArrayCompressor [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+################################################################################
+
+org.nd4j.linalg.cpu.nativecpu.compression.CpuThreshold
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-backend-impls/nd4j-minimizer/src/main/resources/META-INF/services/org.nd4j.linalg.compression.NDArrayCompressor:23]: Any reason why this files has soo many licence headers?

File: nd4j/nd4j-backends/nd4j-backend-impls/nd4j-minimizer/src/main/resources/META-INF/services/org.nd4j.linalg.factory.Nd4jBackend [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+################################################################################
+
+org.nd4j.linalg.minimal.bindings.CpuBackend
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-backend-impls/nd4j-minimizer/src/main/resources/META-INF/services/org.nd4j.linalg.factory.Nd4jBackend:24]: Any reason why this files has soo many licence headers?

**Review by treo — COMMENTED** [review]
Do we really need Both an Openblas delegator and a Blas delegator?

File: nd4j/nd4j-backends/nd4j-backend-impls/nd4j-minimizer/src/main/resources/META-INF/services/org.nd4j.linalg.factory.Nd4jBackend (hunk 4 above)
Comment (agibsonccc) [reply nd4j/nd4j-backends/nd4j-backend-impls/nd4j-minimizer/src/main/resources/META-INF/services/org.nd4j.linalg.factory.Nd4jBackend:24]: I think it just accumulated. I'll work on stripping that down.

File: deeplearning4j/deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/iterator/SparkAMDSI.java (hunk 1 above)
Comment (agibsonccc) [reply deeplearning4j/deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/iterator/SparkAMDSI.java:109]: It's out of date from spark. Somehow this was left over.

File: deeplearning4j/deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/iterator/SparkADSI.java (hunk 2 above)
Comment (agibsonccc) [reply deeplearning4j/deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/iterator/SparkADSI.java:111]: It's out of date from spark. Somehow this was left over.

File: nd4j/nd4j-backends/nd4j-backend-impls/nd4j-minimizer/src/main/resources/META-INF/services/org.nd4j.linalg.compression.NDArrayCompressor (hunk 3 above)
Comment (treo) [reply nd4j/nd4j-backends/nd4j-backend-impls/nd4j-minimizer/src/main/resources/META-INF/services/org.nd4j.linalg.compression.NDArrayCompressor:23]: This still has lots of licence headers.

**Review by treo — CHANGES_REQUESTED** [review]
>contrib/blas-lapack-generator/blas-lapack-generator/OpenblasLapackDelegator.java/org/nd4j/linalg/cpu/nativecpu/OpenblasLapackDelegator.java

>contrib/blas-lapack-generator/blas-lapack-generator/src/main/java/org/nd4j/linalg/cpu/nativecpu/OpenblasLapackDelegator.java

> nd4j/nd4j-backends/nd4j-backend-impls/nd4j-native/src/main/java/org/nd4j/linalg/cpu/nativecpu/OpenblasLapackDelegator.java

Why do we need that file 3 times? It actually looks like the first one is only here by mistake, but why the duplication with the other two?

**Comment by agibsonccc** [comment]
@treo those were just extra generated artifacts. I made sure to delete that now. Thanks for flagging.

**Review by treo — APPROVED** [review]

**Pull request merged**